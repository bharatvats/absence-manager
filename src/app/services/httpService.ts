import { members } from '../../api/json_files/members';
import { absences } from '../../api/json_files/absences';

export const getData = (dataType: string) => {
    return new Promise((resolve) => {
        if (dataType === 'members') {
            resolve(members);
        } else if (dataType === 'absences') {
            resolve(absences);
        }
    })
}