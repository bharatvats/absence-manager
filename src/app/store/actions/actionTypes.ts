export enum AbsenceActionType {
    GetMemberList = 'GET_MEMBER_LIST',
    SetMemberList = 'SET_MEMBER_LIST',
    SetAbsenceList = 'SET_ABSENCES_LIST',
    SetIsLoading = 'SET_IS_LOADING',
    SetIsError = 'SET_IS_ERROR'
}