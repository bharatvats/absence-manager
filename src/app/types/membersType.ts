export enum ResponseStatus {
    SUCCESS = 'Success',
    FAILURE = 'Failure'
}

export interface Member {
    crewId: number;
    id: number;
    image: string;
    name: string;
    userId: number;
}

export interface MembersResponseType {
    message: ResponseStatus;
    payload: Member[];
}

