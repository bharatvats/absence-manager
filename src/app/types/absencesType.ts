import { ResponseStatus } from './membersType';

export enum AbsenceType {
    sickness = 'sickness',
    vacation = 'vacation',
}

export interface Absence {
    admitterId: number | null;
    admitterNote: string;
    confirmedAt: string | null;
    createdAt: string;
    crewId: number;
    endDate: string;
    id: number;
    memberNote: string;
    rejectedAt: string | null;
    startDate: string;
    type: AbsenceType;
    userId: number;
}

export interface AbsencesResponseType {
    message: ResponseStatus;
    payload: Absence[];
}

export enum AbsenceStatus {
    Requested = 'Requested',
    Confirmed = 'Confirmed',
    Rejected = 'Rejected'
}

