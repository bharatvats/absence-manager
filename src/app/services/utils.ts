import { AbsenceType, AbsenceStatus, Absence } from '../types/absencesType';
import { Member } from '../types/membersType';

export interface RowType {
    id: number;
    name?: string;
    type: AbsenceType;
    startDate: number;
    endDate: number;
    period: string;
    memberNote: string;
    status: AbsenceStatus;
    admitterNote: string;
}

export function createRow(absenceEmployee: Absence, member?: Member): RowType {
    return {
        id: absenceEmployee.id,
        name: member?.name,
        type: absenceEmployee.type,
        startDate: new Date(absenceEmployee.startDate).getTime(),
        endDate: new Date(absenceEmployee.endDate).getTime(),
        period: `${absenceEmployee.startDate} ~ ${absenceEmployee.endDate}`,
        memberNote: absenceEmployee.memberNote,
        status: getStatus(absenceEmployee),
        admitterNote: absenceEmployee.admitterNote
    };
}

function getStatus(absenceEmployee: Absence): AbsenceStatus {
    if (absenceEmployee.confirmedAt) {
        return AbsenceStatus.Confirmed;
    } else if (absenceEmployee.rejectedAt) {
        return AbsenceStatus.Rejected;
    } else {
        return AbsenceStatus.Requested;
    }
}