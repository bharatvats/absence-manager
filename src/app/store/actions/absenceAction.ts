import { AbsenceActionType } from './actionTypes';
import { Member } from '../../types/membersType';
import { Absence } from '../../types/absencesType';

export type AbsenceActionReturnTypes = ReturnType<typeof setMemberList>
    | ReturnType<typeof setAbsenceList>
    | ReturnType<typeof setIsLoading>
    | ReturnType<typeof setIsError>;

export function getMemberList(): {
    type: AbsenceActionType.GetMemberList
} {
    return {
        type: AbsenceActionType.GetMemberList
    };
}

export function setMemberList(payload: Member[]): {
    type: AbsenceActionType.SetMemberList,
    payload: Member[]
} {
    return {
        type: AbsenceActionType.SetMemberList,
        payload
    }
}

export function setAbsenceList(payload: Absence[]): {
    type: AbsenceActionType.SetAbsenceList,
    payload: Absence[]
} {
    return {
        type: AbsenceActionType.SetAbsenceList,
        payload
    }
}

export function setIsLoading(isLoading: boolean): {
    type: AbsenceActionType.SetIsLoading,
    payload: boolean
} {
    return {
        type: AbsenceActionType.SetIsLoading,
        payload: isLoading
    }
}

export function setIsError(isError: boolean): {
    type: AbsenceActionType.SetIsError,
    payload: boolean
} {
    return {
        type: AbsenceActionType.SetIsError,
        payload: isError
    }
}