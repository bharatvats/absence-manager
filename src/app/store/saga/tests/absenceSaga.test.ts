import { put, takeLatest, CallEffect, PutEffect } from 'redux-saga/effects';
import { Absence } from '../../../types/absencesType';
import { MembersResponseType, ResponseStatus, Member } from '../../../types/membersType';
import { setAbsenceList, setMemberList, setIsLoading, setIsError } from '../../actions/absenceAction';
import { AbsenceActionType } from '../../actions/actionTypes';
import absenceRootSaga, { getMemberList, getAbsenceList } from '../absenceSaga';
import { mockAbsences, mockMembers } from '../../actions/tests/mockData';
import * as mockAPI from '../../../services/httpService';

describe('Absence saga suite', () => {
    const rootGen = absenceRootSaga();

    it('should wait for latest GetMemberList and execute getMemberList', () => {
        expect(rootGen.next().value)
            .toEqual(takeLatest(AbsenceActionType.GetMemberList, getMemberList))
    });

    it('should be done on next iteration', () => {
        expect(rootGen.next().done).toBeTruthy();
    });

    describe('getMemberList Saga', () => {
        let memberListGen: Generator<CallEffect<unknown> | PutEffect<{
            type: AbsenceActionType.SetMemberList;
            payload: Member[];
        }> | PutEffect<{
            type: AbsenceActionType.SetIsError;
            payload: boolean;
        }>, void, MembersResponseType>

        beforeEach(() => {
            memberListGen = getMemberList();

            const callDescriptor = memberListGen.next().value;
            expect(callDescriptor).toMatchSnapshot();

        });

        it('should call getMemberList API and dispatch success action', async () => {
            const requestMembers = jest.spyOn(mockAPI, 'getData')
                .mockResolvedValue({ "message": ResponseStatus.SUCCESS, payload: mockMembers });


            const putDescriptor = memberListGen.next({ "message": ResponseStatus.SUCCESS, payload: mockMembers }).value
            expect(putDescriptor).toEqual(put(setMemberList(mockMembers)));

            memberListGen.next();

            expect(memberListGen.next().done).toBeTruthy();
            requestMembers.mockClear();
        });

        it('should call getMemberList api and dispatch error action', async () => {
            const requestMembers = jest.spyOn(mockAPI, 'getData')
                .mockResolvedValue({ "message": ResponseStatus.FAILURE, payload: [] });

            const putDescriptor = memberListGen.next({ "message": ResponseStatus.FAILURE, payload: mockMembers }).value
            expect(putDescriptor).toEqual(put(setIsError(true)));

            expect(memberListGen.next().done).toBeTruthy();
            requestMembers.mockClear();
        });
    });

    describe('getAbsenceList Saga', () => {
        let AbsenceListGen: Generator<CallEffect<unknown> | PutEffect<{
            type: AbsenceActionType.SetAbsenceList;
            payload: Absence[];
        }> | PutEffect<{
            type: AbsenceActionType.SetIsError;
            payload: boolean;
        }> | PutEffect<{
            type: AbsenceActionType.SetIsLoading;
            payload: boolean;
        }>>;

        beforeEach(() => {
            AbsenceListGen = getAbsenceList();

            const callDescriptor = AbsenceListGen.next().value;
            expect(callDescriptor).toMatchSnapshot();

        });

        it('should call getMemberList API and dispatch success action', async () => {
            const requestMembers = jest.spyOn(mockAPI, 'getData')
                .mockResolvedValue({ "message": ResponseStatus.SUCCESS, payload: mockAbsences });

            const putDescriptor = AbsenceListGen.next({ "message": ResponseStatus.SUCCESS, payload: mockAbsences }).value;
            expect(putDescriptor).toEqual(put(setAbsenceList(mockAbsences)));

            const loadingDescriptor = AbsenceListGen.next().value;
            expect(loadingDescriptor).toEqual(put(setIsLoading(false)));

            expect(AbsenceListGen.next().done).toBeTruthy();
            requestMembers.mockClear();
        });

        it('should call getMemberList api and dispatch error action', async () => {
            const requestMembers = jest.spyOn(mockAPI, 'getData')
                .mockResolvedValue({ "message": ResponseStatus.FAILURE, payload: mockAbsences });

            const putDescriptor = AbsenceListGen.next({ "message": ResponseStatus.FAILURE, payload: mockAbsences }).value;
            expect(putDescriptor).toEqual(put(setIsError(true)));

            const loadingDescriptor = AbsenceListGen.next().value;
            expect(loadingDescriptor).toEqual(put(setIsLoading(false)));

            expect(AbsenceListGen.next().done).toBeTruthy();
            requestMembers.mockClear();
        });
    });
});