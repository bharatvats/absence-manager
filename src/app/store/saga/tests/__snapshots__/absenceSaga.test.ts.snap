// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Absence saga suite getAbsenceList Saga should call getMemberList API and dispatch success action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "absences",
    ],
    "context": null,
    "fn": [MockFunction],
  },
  "type": "CALL",
}
`;

exports[`Absence saga suite getAbsenceList Saga should call getMemberList api and dispatch error action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "absences",
    ],
    "context": null,
    "fn": [MockFunction],
  },
  "type": "CALL",
}
`;

exports[`Absence saga suite getMemberList Saga should call getMemberList API and dispatch success action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "members",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`Absence saga suite getMemberList Saga should call getMemberList api and dispatch error action 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "members",
    ],
    "context": null,
    "fn": [MockFunction],
  },
  "type": "CALL",
}
`;
